src/finn/custom_op/fpgadataflow/matrixvectoractivation.py

```diff
@@ -165,6 +180,61 @@ def infer_node_datatype(self, model):
         odt = self.get_output_datatype()
         model.set_tensor_datatype(node.output[0], odt)
 
+    def get_input_datatype(self, ind=0):
```

Comment:
There is a lot of code duplication. `get_input_datatype`, `get_weight_datatype`, `get_output_datatype`, ... are all again defined from line 510 onwards. Please remove repetitions.